Make parse-args its own module, add tests

roc opened the pull request:
### behold, the beginning of tests

![screen shot 2016-09-15 at 10 32 43](https://cloud.githubusercontent.com/assets/120181/18545173/dcc741fa-7b2f-11e6-8d02-e09c32304ea9.png)
- tests arguments are passed in as we expect
- raises question about the importance of the `-w` flag
- use unit tested args parser inside hof-transpiler
### Adds homeoffice eslint config, does some linting ✂️  …
- keeping spaces before function parens but most everything else is maintained
- have cheated on the walker nesting errors, because that is work to do later

fixes #19 
begins #16

roc committed:
Make parse-args its own module, add tests

- tests arguments are passed in as we expect
- raises question about the importance of the `-w` flag
- use unit tested args parser inside hof-transpiler

joefitter commented:
@roc we are using https://github.com/UKHomeOffice/eslint-config-homeoffice for linting hof modules - this includes the eslint executable - if using npm >= 3 you can simply add an npm script 

``` json
"lint": "eslint ."
```

if npm < 3 you will need to point to the path: 

``` json
"lint": "./node_modules/eslint-config-homeoffice/node_modules/.bin/eslint ."
```

This provides 2 base .eslintrc files which can be extended for src and test - see here for examples of both: https://github.com/UKHomeOffice/hmpo-enquiries/blob/master/.eslintrc and https://github.com/UKHomeOffice/hmpo-enquiries/blob/master/test/.eslintrc

roc committed:
Add homeoffice eslint config, do some linting ✂️

- keeping spaces before function parens but most everything else is maintained
- have cheated on the walker nesting errors, because that is work to do later

roc commented:
@joefitter added in [38092e8](https://github.com/UKHomeOffice/hof-transpiler/pull/23/commits/38092e82dbf089b0a9781272592db9a4759865f0) but with the odd exception. Not going to try and fix everything in one PR!

JoeChapman commented on .eslintignore:
```diff
@@ -0,0 +1,5 @@
+node_modules
+acceptance_tests
+public
+apps/**/translations/**/default.json
```
don't need this

roc replied:
changed

JoeChapman commented on .eslintignore:
```diff
@@ -0,0 +1,5 @@
+node_modules
+acceptance_tests
+public
```
don't need this

roc replied:
don't need this comment now either

JoeChapman commented on bin/hof-transpiler:
```diff
@@ -1,53 +1,39 @@
+let sources = parsed.sources;
+let shared = parsed.shared;
+
+var mkdirp = require('mkdirp');
```
const?

joefitter replied:
looks like theres a mix of es5 and es6 - lets leave it all as vars until we es6ify or nuke it

joefitter replied:
I mean update the `let`s above to var

roc replied:
`mkdirp` looks like a new line because it's bumped down, but my policy (lol) was to add es6 style variable casting for the pieces I was working on. I'll change them all one way or another depending on what you two decide _in another PR_ (or you could, if it's a bother?)

JoeChapman commented on bin/hof-transpiler:
```diff
@@ -1,53 +1,39 @@
   var writeFileStream = {};
   var firstFileForDir = {};
 
+  /* eslint func-names:0 max-nested-callbacks:0 */
```
maybe try adding a func name and pulling some of the nested functions out?

roc replied:
Out of scope

JoeChapman replied:
ok

JoeChapman commented on lib/parse-args.js:
```diff
@@ -0,0 +1,33 @@
+
+const parse = (args) => {
+  if (!args || args.length === 0) {
+    throw new Error('No src folder or pattern specified');
```
I prefer messages that tell the user how to resolve, maybe "Specify a src folder or pattern"

roc replied:
I'm refactoring to the same spec (same messages). Can improve in another PR

JoeChapman commented on lib/parse-args.js:
```diff
@@ -0,0 +1,33 @@
+
+  if (shared && !Array.isArray(shared)) {
+    shared = [shared];
+  }
```
or just `[].concat(shared)`

roc replied:
no, because `shared` is what's being used, `[].concat(shared)` will be a new array. Unless you mean remove the logic and always `concat`, in which case we'll potentially get `[arr, arr]`

joefitter replied:
yes he means the latter, no it will flatten - [].concat([1, 2, 3]) == [1, 2, 3]

roc replied:
Yes, but if it's already an array then the call will be `[1,2,3].concat([1, 2, 3]) => [ 1, 2, 3, 1, 2, 3 ]`, unless I'm missing something, in which case, I will fall back to saying: another PR

joefitter replied:
why will it already be an array?

joefitter replied:
we are calling `[].concat([1, 2, 3])` or `Array.prototype.concat([1, 2, 3])`

JoeChapman commented on test/lib/parse-args.spec.js:
```diff
@@ -0,0 +1,82 @@
+          result.shared.should.deep.equal([
+            'common/**/bar',
+            'common/**/foo'
+            ])
```
indentation and missing semicolon

roc replied:
I'm not linting tests, feel free to add that in a separate PR

joefitter replied:
this code has been added so should meet our code style guidelines in order to be merged

roc replied:
@joefitter some things on this:
- okay fine
- add a contributing if you're actually serious about this, otherwise 'what guidelines?'
- okay fine

joefitter replied:
sure, we can add a contributing file, but badly formatted code (indentation) should go without saying

JoeChapman commented on test/lib/parse-args.spec.js:
```diff
@@ -0,0 +1,82 @@
+      });
+      describe('with --writeShared', () => {
+
+        it('makes no actual difference to entire thing')
```
what?

roc replied:
It doesn't, I'm fairly sure we still get a compiled file anyway, and I don't think you can change the destination correctly as is. I would be happy if any other person fancied writing a single other test to disprove

JoeChapman replied:
ok, but couldn't you remove this line? If you want to raise awareness an issue would be cool.

joefitter replied:
writeShared writes the output to the shared directory. It makes no difference to the target directories, but if you are using shared it does (or did) make a difference

joefitter replied:
if it is broken please raise an issue

roc replied:
I'm not sure it's broken, I'm a confused person adding tests, I can raise an issue re: this though if you like

joefitter replied:
I'd just remove this test for now then

JoeChapman commented on test/lib/parse-args.spec.js:
```diff
@@ -0,0 +1,82 @@
+          result.shared.should.deep.equal([
+            'common/**/bar',
+            'common/**/foo'
+          ])
```
missing semicolon

roc replied:
not linting tests

JoeChapman replied:
We'd like it if you would

```
extends:
  - '../.eslintrc'
  - 'homeoffice/config/testing'
```

roc replied:
fine

JoeChapman commented on test/lib/parse-args.spec.js:
```diff
@@ -0,0 +1,82 @@
+        should.not.exist(result.shared);
+      });
+
+      it('throws with array')
```
what?

JoeChapman commented on test/lib/parse-args.spec.js:
```diff
@@ -0,0 +1,82 @@
+    [[],Number(),'',null,undefined].forEach((type) => {
+      it(`${typeof type} '${type}' should throw`, () => {
+        (() => {
+          parse(type)
```
semicolon

JoeChapman commented on test/globals.js:
```diff
@@ -0,0 +1,6 @@
+global.chai = require('chai')
+global.should = global.chai.should()
+global.sinon = require('sinon')
+global.proxyquire = require('proxyquire')
```
missing semicolons

joefitter replied:
do we even want these to be global? Why dont we just require them when we need them?

roc replied:
I find it useful for removing test boilerplate, take them out in a subsequent PR that shows how useful it is not to globalise them if you like

JoeChapman replied:
I think it's typical of our projects @joefitter, but we don't usually make proxyquire gobal.

roc replied:
Have removed sinon and proxyquire references entirely, as we're not using them yet https://github.com/UKHomeOffice/hof-transpiler/pull/23/commits/9c77d76409b280d2e08972ac9174a78634a032ad

roc committed:
Remove unnecessary references

- Removes references to non-existent files from .eslintignore
  - we don't have acceptance_tests / public / apps folders

- Removes unneccessary `proxyquire` and `sinon` modules
  - not using these modules (yet)

roc commented on test/lib/parse-args.spec.js:
```diff
@@ -0,0 +1,82 @@
+          result.shared.should.deep.equal([
+            'common/**/bar',
+            'common/**/foo'
+          ])
```
UGH

roc committed:
Add linting to tests, run lint --fix
